File: src/easy/Q0020.py
"""
20. Valid Parentheses
https://leetcode.com/problems/valid-parentheses/
"""


# Time: O(n), Space: O(n), where n is the length of `s`
def is_valid(s: str):
    pairs = {")": "(", "}": "{", "]": "["}
    stack = []
    for bracket in s:
        if bracket in pairs.values():
            stack.append(bracket)
            continue
        if not stack:
            return False
        if stack.pop() != pairs[bracket]:
            return False
    return len(stack) == 0

File: src/main/java/easy/Q0020.java
package easy;

import java.util.Map;
import java.util.Objects;
import java.util.Stack;

// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/
public class Q0020 {

    // Time: O(n), Space: O(n), where n is the length of s.
    boolean isValid(String s) {
        var brackets = Map.of(
                '(', ')',
                '{', '}',
                '[', ']'
        );
        var open = new Stack<Character>();
        for (var c : s.toCharArray()) {
            if (brackets.containsKey(c)) {
                open.push(c);
                continue;
            }
            if (open.empty()) {
                return false;
            }
            var top = open.pop();
            if (!Objects.equals(c, brackets.get(top))) {
                return false;
            }
        }
        return open.empty();
    }
}
